File: apps/ruby/design-pattern/14-factory/test/factory_test.rb
require_relative "test_helper"
require_relative "../lib/organisms"
require_relative "../lib/pond"
require_relative "../lib/organism_factory"
require_relative "../lib/habitat"

class DuckPondTest < Minitest::Test
  def test_simulate_one_day_includes_duck_and_water_lily_actions
    pond = DuckPond.new(number_animals: 2, number_plants: 1)
    output = pond.simulate_one_day

    assert_includes output, "WaterLily is growing."
    assert_includes output, "Quack!"
    assert_includes output, "Duck is eating."
    assert_includes output, "Duck is sleeping."
  end

  def test_correct_number_of_organisms_are_simulated
    pond = DuckPond.new(number_animals: 3, number_plants: 2)
    output = pond.simulate_one_day

    assert_equal 2, output.count("WaterLily is growing.")
    assert_equal 3, output.count("Quack!")
  end
end

class HabitatWithOrganismFactoryTest < Minitest::Test
  def test_jungle_habitat_simulates_tree_and_tiger_actions
    jungle_factory = OrganismFactory.new(Tree, Tiger)
    jungle = Habitat.new(2, 3, jungle_factory)
    output = jungle.simulate_one_day

    assert_equal 3, output.count("Tree is growing.")
    assert_equal 2, output.count("Roar!")
    assert_includes output, "Tiger is eating."
  end

  def test_organism_factory_creates_correct_types
    factory = OrganismFactory.new(Algae, Duck)

    assert_instance_of Duck, factory.new_animal
    assert_instance_of Algae, factory.new_plant
  end
end

File: apps/java/design-pattern/src/test/java/pattern/factory/FactoryTest.java
package pattern.factory;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class FactoryTest {

    @Test
    void duckBehavior() {
        Animal duck = new Duck();
        assertEquals("Quack!", duck.speak());
        assertEquals("Duck is eating.", duck.eat());
        assertEquals("Duck is sleeping.", duck.sleep());
    }

    @Test
    void duckPondSimulatesOneDay() {
        Pond pond = new DuckPond(2, 1);
        List<String> output = pond.simulateOneDay();

        assertTrue(output.contains("WaterLily is growing."));
        assertTrue(output.contains("Quack!"));
        assertTrue(output.contains("Duck is eating."));
        assertTrue(output.contains("Duck is sleeping."));
    }

    @Test
    void frogPondSimulatesOneDay() {
        Pond pond = new FrogPond(1, 1);
        List<String> output = pond.simulateOneDay();

        assertTrue(output.contains("Algae is growing."));
        assertTrue(output.contains("Croak!"));
        assertTrue(output.contains("Frog is eating."));
    }

    @Test
    void abstractFactoryCreatesOrganisms() {
        OrganismFactory factory = new OrganismFactory(Tiger::new, Tree::new);
        Animal animal = factory.newAnimal();
        Plant plant = factory.newPlant();

        assertEquals("Roar!", animal.speak());
        assertEquals("Tree is growing.", plant.grow());
    }

    @Test
    void habitatUsesAbstractFactory() {
        OrganismFactory factory = new OrganismFactory(Duck::new, WaterLily::new);
        Habitat habitat = new Habitat(2, 1, factory);
        List<String> output = habitat.simulateOneDay();

        assertTrue(output.contains("WaterLily is growing."));
        assertTrue(output.contains("Quack!"));
        assertEquals(7, output.size()); // 1 plant + 2 animals * 3 actions
    }

    @Test
    void tigerSpeaksRoar() {
        Animal tiger = new Tiger();
        assertEquals("Roar!", tiger.speak());
        assertEquals("Tiger is eating.", tiger.eat());
        assertEquals("Tiger is sleeping.", tiger.sleep());
    }
}
